File: continuous_subarr_sum.py
# 523. Continuous Subarray Sum


class Solution(object):
    def checkSubarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        dic, sum = {0: -1}, 0
        for i, n in enumerate(nums):
            if k != 0:
                sum = (sum + n) % k
            else:
                sum += n
            if sum not in dic:
                dic[sum] = i
            else:
                if i - dic[sum] >= 2:
                    return True
        return False


if __name__ == "__main__":
    obj = Solution()
    print(obj.checkSubarraySum(nums=[23, 2, 4, 6, 7], k=6))
    print(obj.checkSubarraySum(nums=[23, 2, 6, 4, 7], k=6))
    print(obj.checkSubarraySum(nums=[23, 2, 6, 4, 7], k=13))

File: other/continuous_subarr_sum.cpp
#include "leetcode.h"

class Solution {
public:
  bool checkSubarraySum(vector<int> &nums, int k) {
    unordered_map<int, int> um;
    int s = 0;
    um[0] = -1;
    for (int i = 0; i < nums.size(); i++) {
      s += nums[i];
      if (k != 0)
        s %= k;
      if (um.find(s) != um.end()) {
        if (i - um[s] > 1)
          return true;
      } else
        um[s] = i;
    }
    return false;
  }
};

int main() {
  Solution obj;
  vector<int> nums = {23, 2, 4, 6, 7};
  int k = 6;
  if (obj.checkSubarraySum(nums, k))
    cout << "true";
  else
    cout << "false";
}
